dcc-storage-server/src/main/java/org/icgc/dcc/storage/server/service/upload/UploadStateStore.java

```diff
@@ -278,18 +290,21 @@ public void finalizeUploadPart(String objectId, String uploadId, int partNumber,
   }
 
   @SneakyThrows
-  public List<PartETag> getUploadStatePartEtags(String objectId, String uploadId) {
+  public Map<Integer, UploadPartDetail> getUploadStatePartDetails(String objectId, String uploadId) {
     val uploadStateKey = getUploadStateKey(objectId, uploadId, PART);
-    val etags = Lists.<PartETag> newArrayList();
+    val details = Maps.<Integer, UploadPartDetail> newHashMap();
 
     eachObjectSummary(objectId, uploadStateKey, (objectSummary) -> {
       CompletedPart part = readCompletedPart(objectId, uploadId, objectSummary);
+      UploadPartDetail detail = new UploadPartDetail();
       PartETag etag = new PartETag(part.getPartNumber(), part.getEtag());
-
-      etags.add(etag);
+      detail.setEtag(etag);
```

Comment:
Ya `@Value` with `@Builder` seems more appropriate. Better thread safety and invariant guarantees.